[Localization] Set the correct currency when switching locales

adamraya opened the pull request:
GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000005MF7MYAW/view

# Description

This PR implements the specific association of locales and currencies represented in the table for all the Commerce API requests that accept the `currency` parameter.

| Locale  | Currency |
| ------------- | ------------- |
| en-GB  | GBP |
| fr-FR  | EUR |
| it-IT  | EUR |
| zh-CN | CNY |
| ja-JP  | JPY |

The preferred currencies for each langue are defined in the `package.json` file:
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/a1ba9329d7dbec1aab5e947d508e50d97a6217cc/packages/pwa/package.json#L186-L217


The PR also refactors:
- Categories context Provider by adding a new `<CategoriesProvider />` and a new `useCategory()` hook.
- The util function `buildUrlLocale` now called `getUrlWithLocale`.
---
These are some of the API calls that include the `currency` value as a query parameter or in the body of the request:

## PLP
**PENDING SCAPI FIX**: The `currency` query parameter sent to the `product-search` endopint is ignored. The response always uses the default currency `"currency": "GBP"`. [Slack thread](https://salesforce-internal.slack.com/archives/C02B75CCFV2/p1631300673055200).

#### SCAPI [`ShopperSearch`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shoppersearch.shoppersearch-1.html) 
Methods: [`productsearch`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shoppersearch.shoppersearch-1.html#productsearch).

## PDP
#### SCAPI [`ShopperProducts`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html) 
Methods: [`getProduct`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html#getproduct) and [`getProducts`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html#getproducts).

## Cart/Checkout/Add to cart
The Basket is created using the method `createBasket` of the `OcapiShopperBaskets` class.

#### OCAPI [`Baskets`](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html) resource
| HTTP method  | Resource |
| ------------- | ------------- |
| POST  | [/baskets](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html#iid-1036386844__id49484514) |

The POST request creates a new basket initialized with default values. This method apparently doesn't allow us to customize the `currency` value on the body of the request.

I added a new method `updateBasket` to the `OcapiShopperBaskets` class that updates the basket currency.
The method is called when the current currency doesn't match the current locale.
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/2ba4f636eda0ff9da05f19ae129c82f785abccc9/packages/pwa/app/commerce-api/ocapi-shopper-baskets.js#L28

The `updateBasket` method uses the OCAPI:
#### OCAPI [`Baskets`](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html) resource
| HTTP method  | Resource |
| ------------- | ------------- |
| PATCH  | [/baskets/{basket_id}](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html#id-1036386844__id1441479317) |



<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relavant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [X] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes
- Added currencies for each locale.
- Added currency parameter to API calls.
- Added `updateBasket` method to the `OcapiShopperBaskets` class to updates the basket currency.
- Refactor `useCategory` and `useCurrency` hooks and contexts providers `<CategoriesProvider />` and `<CurrencyProvider />`.
- Fixed and added tests

# How to Test-Drive This PR

1. Navigate across all the pages that show a currency value (Home, PLP, PDP, Cart, Checkout, My account, etc.)
2. Verify that the currency matches the selected locale as per the association of locales and currencies in the table.



<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [X] Changes are covered by test cases
- [X] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [X] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

adamraya committed:
Use currency parameter for getProduct on PDP

adamraya committed:
Clean up

adamraya committed:
Update basket currency

adamraya committed:
Use the default currency associated to the default locale

adamraya committed:
Fix tests

bendvc commented on packages/pwa/app/commerce-api/hooks/useBasket.js:
```diff
@@ -54,18 +54,50 @@ export default function useBasket() {
 
                 if (Array.isArray(customerBaskets?.baskets)) {
-                    return setBasket(customerBaskets.baskets[0])
+                    let result
+
+                    result = setBasket(customerBaskets.baskets[0])
+
+                    // Update basket currency if necessary
+                    result =
+                        currency &&
+                        customerBaskets.baskets[0].currency !== currency &&
+                        this.updateBasketCurrency(currency, customerBaskets.baskets[0].basketId)
+
+                    return result
```
There is a point of failure in this logic. In the scenario where you call `getOrCreateBasket()`, notice I didn't pass in a currency value, you'll make a request to get the customers baskets, set the state to be the first basket and then return result which is now assigned to `undefined`.

There are a couple things you can do here.

1. You can make currency a required param and throw and exception if it's undefined.
2. Change the signature of this function to include currency inside an options/opts param. Then clean up the logic as follows.


```
            async getOrCreateBasket(opts = {}) {
                const {currency} = opts
                const customerBaskets = await api.shopperCustomers.getCustomerBaskets({
                    parameters: {customerId: customer?.customerId}
                })
                
                // Throw if there was a problem getting the customer baskets
                isError(customerBaskets)) &&
                    throw new Error(customerBaskets)
                
                // We only support single baskets for now. Grab the first one. This will probably work as long as baskets is an array.. we might have to use `.?[0]`
                let basket = customerBaskets.baskets[0]
                
                if (!basket) {
                    // NOTE: Can we pass the currency in there?
                    basket = await api.shopperBaskets.createBasket({})
                    // NOTE: basket could be an error, you'll want to address that.
                    setBasket(basket)
                }
                
                 // Update basket currency if it was created with the wrong one, this will also set the state.
                if (basket.currency !== currency) {
                    this.updateBasketCurrency(currency, newBasket.basketId)
                }

                return basket
            },
```

Obviously I think the later is the better approach because we might support multiple baskets in the future, so you should see use passing more options like `id` to get a basket with a specific id, and the api doesn't require currency for getting or creating.

bendvc replied:
Update, maybe we put the options on the hook, and not the individual function. This way `currency` will be available to all the function defined in the hook.

bendvc commented on packages/pwa/app/commerce-api/hooks/useBasket.js:
```diff
@@ -54,18 +54,50 @@ export default function useBasket() {
+                    body: {currency},
+                    parameters: {basketId}
+                })
+                if (updateBasket.fault) {
```
There is a `isError` util that we use in other hooks, maybe you want to see if that will work here. For consistency sake.

bendvc commented on packages/pwa/app/commerce-api/hooks/useShopper.js:
```diff
@@ -15,11 +15,10 @@ import {customerProductListTypes} from '../../constants'
  * @returns {Object} - customer and basket objects
  */
-const useShopper = () => {
+const useShopper = (currency) => {
```
Similar to how `currency` was optional in `getOrCreateBasket` we should probably do that here. It will work without it right?

Thinking about it a little more, maybe `getOrCreateBasket` shouldn't have the options object as descripted above. But pass the options to the hook like this:


```
const basket = useBasket({currency: 'USD'})
```

And the `getOrCreateBasket` will have access to it from its scope.

bendvc replied:
Doing it this way will reduce how often you have to pass around the `currency` like in the 2 calls below.

adamraya committed:
PR Feedback clean up getOrCreateBasket

adamraya committed:
PR Feedback currency as options on the hooks

adamraya committed:
Set basket state only once on getOrCreateBasket

adamraya committed:
Merge branch 'develop' into W-9871843_set_correct_currency

adamraya committed:
Increase bundle size

vmarta commented on packages/pwa/package.json:
```diff
@@ -184,11 +184,26 @@
-            "it-IT",
-            "zh-CN",
-            "ja-JP"
+            {
+                "id": "en-GB",
+                "currencyCode": "GBP"
+            },
```
Do we need to update [this readme](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/17405cf6b86b6581e9c9c7ae61534b8e973c57ca/packages/pwa/app/translations/README.md) too?

adamraya committed:
Use preferredCurrency instead of currencyCode

bendvc commented on packages/pwa/app/components/cart-item-variant/item-price.jsx:
```diff
@@ -39,7 +40,7 @@ const ItemPrice = ({currency, ...props}) => {
                 <FormattedNumber
                     style="currency"
-                    currency={currency || basket.currency || DEFAULT_CURRENCY}
+                    currency={currency || basket.currency || getCurrency(DEFAULT_LOCALE)}
```
You can probably dry this up, assign a variable for this derived currency. Same goes for line 57 and 68

bendvc commented on packages/pwa/app/components/footer/index.jsx:
```diff
@@ -135,8 +135,8 @@ const Footer = ({...otherProps}) => {
                                 {SUPPORTED_LOCALES.map((locale) => (
-                                    <option key={locale} value={locale}>
-                                        {intl.formatMessage(defaultLocaleMessages[locale])}
+                                    <option key={locale.id} value={locale.id}>
```
Try simulating a slow network and you'll see that after selecting a new locale from the dropdown you won't see it being selected. To fix this you can use a local state for the locale value like this:
```
const [locale, setLocale] = useState(intl.locale)
```

then use the locale value as the select value, and also setLocale(target.value) in the onChange function.

adamraya committed:
PR Feedback

adamraya committed:
PR Feedback update translations README.md

adamraya committed:
Merge branch 'develop' into W-9871843_set_correct_currency

# Conflicts:
#	packages/pwa/app/components/cart-item-variant/item-price.jsx

bendvc commented on packages/pwa/package.json:
```diff
@@ -183,12 +183,33 @@
         "last 4 ChromeAndroid versions"
     ],
     "l10n": {
+        "supportedCurrencies": [
```
I like the attention to detail on how to reordered those properties :)

bendvc commented on packages/pwa/app/utils/url.js:
```diff
@@ -114,7 +114,9 @@ export const buildUrlLocale = (previousLocale, newLocale) => {
         ? homeUrlBuilder(HOME_HREF, newLocale)
         : `/${
-              SUPPORTED_LOCALES.includes(newLocale) ? newLocale : DEFAULT_LOCALE
+              SUPPORTED_LOCALES.find((locale) => locale.id === newLocale)
```
I wouldn't feel right if I didn't comment on this function: 

1. The name isn't descriptive, after reading it I'd think it builds some kind of locale. Maybe something like getCurrentUrlWithLocale?
2. The fact that we are singling out `refine` as a param to remove is odd.

Here is a partial implementation that I feel reads and has a better api:

```


const getUrlWithLocale = (shortCode, opts = {}) => {
    const {location} = window
    const {disallowParams = []} = opts
    let relativeUrl = `${location.pathname}${location.search}`

    // Replace the locale short-code.
    relativeUrl = relativeUrl.replace(SUPPORTED_LOCALES_REGEX, `/${shortCode}`)

    // Remove any disallowed params.
    if (disallowParams.length) {
        const params = new URLSearchParams(location.search)
        disallowParams.forEach((param) => {
            params.delete(param)
        })

        // TODO: replace older search with new search values
    }

    return relativeUrl
}

```

adamraya committed:
Add updateBasket tests

bendvc commented on packages/pwa/app/components/_app-config/index.jsx:
```diff
@@ -76,7 +77,9 @@ const AppConfig = ({children, locals = {}}) => {
-    locals.api = new CommerceAPI({...apiConfig, locale})
+    const currency = getCurrency(locale)
+
+    locals.api = new CommerceAPI({...apiConfig, locale, currency})
```
We previously made a request to the `commerce-sdk-isomorphic` team to support `locale` as a top-level configuration value for the api's. I don't see any reason not to make the same request to have `currency` also be a top-level config value. I'm going to leave a message for Ujwala Nandyala who's the PM on that team now. To see if we can get that on their radar.

vmarta commented on packages/pwa/app/components/add-to-cart-modal/index.jsx:
```diff
@@ -45,6 +46,7 @@ const AddToCartModal = ({product, variant, quantity, isOpen, onClose, ...props})
         viewType: 'small',
         selectedVariationAttributes: variationValues
     })?.images?.[0]
+    const defaultLocaleCurrency = getCurrency(DEFAULT_LOCALE)
```
It looks like we call `getCurrency(DEFAULT_LOCALE)` in several different files. To avoid repeating it, maybe we can implement `getDefaultCurrency()` in locale.js. Or perhaps have `const DEFAULT_CURRENCY = getCurrency(DEFAULT_LOCALE)` set in constants.js.

adamraya replied:
Yeah, I agree, it wasn't the best approach.
Now we use the `useCurrency()` hook.

vmarta commented on packages/pwa/app/utils/locale.js:
```diff
@@ -90,3 +92,17 @@ export const getTargetLocale = (preferredLocales, supportedLocales, defaultLocal
+ * @param locale The locale that we want the currency
+ * @returns {string} Either the currency specific for the locale OR the currency for the default locale
+ */
+export const getCurrency = (locale) => {
```
Minor: do we want to test this simple function? Oh doing so may help pass the coverage tests.

adamraya replied:
True, I added a few tests.

adamraya committed:
Refactor getUrlWithLocale util function

adamraya committed:
Add tests

adamraya committed:
Add CurrencyContext

bendvc committed:
Refactor CatgoriesProvider and CurrencyProvider

make them look more familiar, like IntlProvider, and more consistent in how they are designed.

adamraya committed:
Add getCurrency tests & Docs & Changelog

adamraya committed:
Add hooks tests

adamraya committed:
Merge branch 'develop' into W-9871843_set_correct_currency

# Conflicts:
#	packages/pwa/CHANGELOG.md
#	packages/pwa/app/commerce-api/hooks/useBasket.js
#	packages/pwa/app/commerce-api/hooks/useShopper.js
#	packages/pwa/app/pages/product-list/index.jsx
#	packages/pwa/app/utils/test-utils.js
#	packages/pwa/package-lock.json

adamraya committed:
Solve merge conflicts

adamraya committed:
Update translations README.md

bendvc commented on packages/pwa/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## v1.2.0 (Month Day, 2021)
```
Typically we label this as "To be released" versus labeling a version and a placeholder for dates.

adamraya committed:
Update CHANGELOG.md

vmarta commented on packages/pwa/app/hooks/use-categories.js:
```diff
@@ -0,0 +1,14 @@
+
+/**
+ * Custom React hook to get the categories
+ * @returns {{currency: string, setCurrency: function}[]}
```
Oh looks like this is a typo due to copy & paste..

```suggestion
 * @returns {{categories: Object, setCategories: function}[]}
```

vmarta commented on packages/pwa/app/utils/url.test.js:
```diff
@@ -94,6 +97,71 @@ describe('url builder test', () => {
+    })
 })
 
 describe('rebuildPathWithParams test', () => {
```
I believe at this point, the value for `window.location` is _not_ going to be the original one. Perhaps we can use `beforeEach`/`afterEach` to make sure that location is reset properly.

adamraya replied:
good catch! a typical mistake that causes effects in other tests. I'm using now `beforeEach`/`afterEach` to restore the original `window.location`.

vmarta commented on packages/pwa/app/utils/url.js:
```diff
@@ -102,24 +102,40 @@ export const productUrlBuilder = (product) => encodeURI(`/product/${product.id}`
+ * @returns {string} - The relative URL for the specific locale.
  */
-export const buildUrlLocale = (previousLocale, newLocale) => {
+export const getUrlWithLocale = (shortCode, opts = {}) => {
```
Can you extract changes related to `getUrlWithLocale` into a separate PR perhaps? I think they can stand on their own, and I'll need them too in my localization work: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/137/files#diff-f46ab45bbea466e88ac5ab4c430662f8f265bce3754178d40040643bdfc72b3aR113

adamraya replied:
The currency bug in the SCAPI `product-search` endpoint is fixed now, so we should be good to merge this PR.

adamraya committed:
Update packages/pwa/app/hooks/use-categories.js

Co-authored-by: Vincent Marta <vmarta@gmail.com>

adamraya committed:
PR Feedback

adamraya committed:
Merge branch 'develop' into W-9871843_set_correct_currency

bendvc committed:
Make the code for sending currency and locale params more flexible and readable

bendvc commented on packages/pwa/app/components/_app-config/index.jsx:
```diff
@@ -40,7 +41,7 @@ const getLocale = (locals) => {
     const localeUrl = originalUrl && originalUrl.split('/')[1]
```
This code is not solid. If you goto the url `http://localhost:3000/it-IT?site=broken` you'd expect to see the site in italian, but you see it in the default locale because of this split isn't really the smartest code. You might have to first remove the query string if there is one.

bendvc committed:
Merge branch 'develop' into W-9871843_set_correct_currency

bendvc committed:
Fix getLocale function

bendvc committed:
Rename getCurrency to getPreferredCurrency

- This is to avoid confusion with the current active currency as provided by the provider, and that which is just the preferred currency for a given locale
